### PR TITLE
feat: support facet ordering

### PR DESCRIPTION
**Pull request by TomKlotzPro**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes    
| BC breaks?        | no     
| Related Issue     | [162](https://github.com/algolia/algoliasearch-client-specs-internal/issues/163)
| Need Doc update   | yes

## Describe your change

The new Facet Ordering feature providing the order of facet widgets and facet values.

### CODE REVIEW

File: src/Algolia.Search/Models/Settings/IndexSettings.cs [hunk 1]
```diff
@@ -91,6 +92,13 @@ private List<string> AttributesToIndex
         /// </summary>
         public int? RelevancyStrictness { get; set; }
 
+        /// <summary>
+        /// Content defining how the search interface should be rendered. 
+        /// This is set via the settings for a default value and can be overridden via rules.
+        /// </summary>
+        [JsonExtensionData]
```
Comment (Ant-hem) [inline src/Algolia.Search/Models/Settings/IndexSettings.cs]:
```suggestion
```

I don' think that's useful, is that a glyph? see 86bcb969249c21198b07d540362aa754b97bf5e0

**Review by Ant-hem — COMMENTED** [review]
LGTM. I pushed some modifications to speed-up a bit the review cycle. 
Can you also review my changes @TomKlotzPro and @VladislavFitz?

note: I'd have add some serialization/deserialization tests but I don't know what's the current testing strategy for API-Clients.

File: src/Algolia.Search/Models/Rules/FacetOrdering.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+{
+    /// <summary>
+    /// Facets and facets values ordering rules container.
+    /// </summary>
+    public class FacetOrdering
+    {
+        /// <summary>
+        /// Define or override the way facet attributes are displayed.
+        /// </summary>
+        public FacetsOrder Facets { get; set; }
```
Comment (Haroenv) [inline src/Algolia.Search/Models/Rules/FacetOrdering.cs:36]: these are optional, how is that defined in c#?

File: src/Algolia.Search/Models/Rules/FacetValuesOrder.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        /// <summary>
+        /// An ordered list of facet values to present.
+        /// </summary>
+        public List<string> Order { get; set; }
+
+        /// <summary>
+        /// A constant defining what to do with remaining facet values not included in FacetsOrder.
+        /// </summary>
+        public SortRemainingByType SortRemainingBy { get; set; }
```
Comment (Haroenv) [inline src/Algolia.Search/Models/Rules/FacetValuesOrder.cs:42]: also optional

File: src/Algolia.Search/Models/Settings/IndexSettings.cs [hunk 4]
```diff
@@ -85,12 +86,18 @@ private List<string> AttributesToIndex
...
+        /// Bigger value means less, but more relevant results,
         /// lesser value - more less relevant results
         /// </summary>
         public int? RelevancyStrictness { get; set; }
 
+        /// <summary>
+        /// Content defining how the search interface should be rendered.
+        /// This is set via the settings for a default value and can be overridden via rules.
+        /// </summary>
+        public IDictionary<string, object> RenderingContent { get; set; }
```
Comment (Haroenv) [inline src/Algolia.Search/Models/Settings/IndexSettings.cs:99]:
```suggestion
        public RenderingContent RenderingContent { get; set; }
```

is this possible?

File: src/Algolia.Search/Models/Rules/FacetOrdering.cs (hunk 2 above)
Comment (VladislavFitz) [reply src/Algolia.Search/Models/Rules/FacetOrdering.cs:36]: If I understood correctly, reference types such as class instances are nullable (optional) by definition, so no need to declare them as optional explicitly. Could you confirm @Ant-hem ?

**Review by VladislavFitz — APPROVED** [review]

File: src/Algolia.Search/Models/Rules/FacetOrdering.cs (hunk 2 above)
Comment (sbellone) [reply src/Algolia.Search/Models/Rules/FacetOrdering.cs:36]: That's right @VladislavFitz, not present = null = not serialized.

**Review by Ant-hem — APPROVED** [review]

**Pull request merged**